cleanup PK parse - part 1

mpg opened the pull request:
## Description

This is part 1 of  #7994 - focusing on  `MBEDTLS_PK_USE_PSA_EC_DATA` mostly, plus miscellaneous other things.

I decided to stop after about 20 commits in order to keep it reviewable, I'll continue #7994 in another PR once this one is OK.

Note to reviewers: some commits just move code around and can easily be reviewed with `git show --color-moved <commit_id>`.

## PR checklist

- [x] **changelog** not required - internal refactoring
- [x] **backport** not required - refactoring only applies to 3.x
- [x] **tests** not required - covered by existing tests

valeriosetti commented on library/pkparse.c:
```diff
@@ -53,177 +55,341 @@
+ * Note: the private key information is always available from pk,
+ * however for convenience the serialized version is also passed,
+ * as it's available at each calling site, and useful in some configs
+ * (as otherwise we're have to re-serialize it from the pk context).
```
```suggestion
 * (as otherwise we would have to re-serialize it from the pk context).
```

valeriosetti commented on library/pkparse.c:
```diff
@@ -53,177 +55,341 @@
+    mbedtls_ecp_keypair_init(&ecp_key);
+    ret = mbedtls_ecp_group_load(&(ecp_key.grp), ecp_group_id);
+    if (ret != 0) {
         return ret;
```
`ecp_key` is already initialized here, so I think we need to `goto exit;` in case of failure instead of returning. Isn't it?

valeriosetti commented on library/pkparse.c:
```diff
@@ -53,177 +55,341 @@
+ * Tell if the passed tag might be the start of SpecifiedECDomain
+ * (that is, a sequence).
+ */
+static int pk_ecc_tag_may_be_specified_ec_domain(int tag)
```
For these kind of functions returning a sort of yes/no answer, I usually prefer `is` in the name. So it would become `pk_ecc_tag_is_specified_ec_domain`. However this is really personal, so if you prefer it that way I'm OK with that ;)

mpg replied:
Agreed, see the commit message for why I originally went with "may be" but now think those reasons were wrong :)

valeriosetti reviewed (commented):
I only have a couple of comments but the overall shape is looking really good!
I think there is still something to be done as there is a commend on line 635 stating that there is a section which is still `Unsorted (yet!) from this point on until the next section header`. I will take another look once this is done ;)

mpg commented:
>I think there is still something to be done as there is a commend on line 635 stating that there is a section which is still Unsorted (yet!) from this point on until the next section header. I will take another look once this is done ;)

My plan was to address that in a follow-up PR, in order to avoid this one getting too big: 22 commits is already not that small, and there may be more coming from reviews and the CI that is not fully passing yet.

mpg commented:
@valeriosetti Thanks for your review. I've addressed your comments, and fixed the first two errors reported by the CI - let's see if that was everything or if there as still failures in the next run.

valeriosetti reviewed (approved):
Thanks for addressing my comments. LGTM.
Nice work! ;)

valeriosetti reviewed (dismissed):
LGTM. I hope that the CI agrees as well :)

valeriosetti reviewed (dismissed):
LGTM ;)

d3zd3z commented on library/pkparse.c:
```diff
@@ -50,174 +55,346 @@
-int mbedtls_pk_parse_keyfile(mbedtls_pk_context *ctx,
-                             const char *path, const char *pwd,
-                             int (*f_rng)(void *, unsigned char *, size_t), void *p_rng)
+static int pk_ecc_set_pubkey_from_prv(mbedtls_pk_context *pk,
```
With this now distinguishing between invalid and unsupported, do we need to add a test to test this distinction?

mpg replied:
I'm sorry, I don't understand. Can you clarify "this now distinguishing between invalid and unsupported"?

d3zd3z commented on library/pkparse.c:
```diff
@@ -50,174 +55,346 @@
     }
-
     return 0;
+#endif /* MBEDTLS_PK_USE_PSA_EC_DATA */
```
This comment is a little confusing, shouldn't it be something like `#endif /* not MBEDTLS_PK_USE_PSA_EC_DATA */` to indicate the logic of the code above it. Not sure what the best comment on the else is.

mpg replied:
Well, that's how we do it in most places: the comment just repeats the contents of the `#if` regardless of whether there's an `#else` in the middle or not. I agree it's debatable, but that's our current style.

d3zd3z reviewed (approved):
Two little things, not really worthy of keeping the change from going in, but things to consider maybe in future changes.
